concepts.tex

```diff
@@ -803,6 +564,321 @@
 \end{itemize}
 \end{itemdescr}
 
+\rSec1[concepts.lib.object]{Object concepts}
+
+\pnum
+This section \oldtxt{describes the foundational}\newtxt{defines} concepts that describe the basis of the
+value-oriented programming style on which the library is based. \oldtxt{The purpose of these concepts is
+to establish a foundation for equational reasoning in programs.} \ednote{These concepts reuse many of the
+names of concepts that traditionally describe features of types to describe families of object types (See
+the rationale in Appendix~\ref{decomposition}).}
+
+\rSec2[concepts.lib.object.destructible]{Concept Destructible}
+\ednote{Remove table [destructible] in [utility.arg.requirements]. Replace
+references to [destructible] with references to
+[concepts.lib.\oldtxt{corelang}\newtxt{object}.destructible].}
+
+\pnum
+\newtxt{The \tcode{Destructible} concept is the base of the hierarchy of object concepts.
+It describes properties that all such object types have in common.}
+
+\indexlibrary{\idxcode{Destructible}}%
+\begin{itemdecl}
+template <class T>
+concept bool Destructible@\newtxt{() \{}\oldtxt{ =}@
+  @\newtxt{return}@ requires (T t@\newtxt{, T* p}@) {
+    { t.@$\sim$@T() } noexcept;
+    @\newtxt{\{ \&t \} -> Same<T*>;}@
+    @\newtxt{delete p;}@
+    @\newtxt{delete[] p;}@
+  };
+@\newtxt{\}}@
+\end{itemdecl}
```

Comment:
I've considered it, and ended up in the position of "wait and see." I hate to weaken the concepts without a motivating need, especially disjunctions of unnamed precursor concepts. If we define a concept C = A || B where A and B are internal/unnamed and someone actually cares about the difference between A and B they have to find a way to discriminate the two. If we instead provide A and B as named concepts that problem doesn't exist, and uses that don't need to discriminate can easily form the disjunction of A and B.

It's analagous to my view of type erasure. Yes, type erasure is useful, but in general it's a good idea to push the erasure out toward the edges of the design since it's hard to unerase types.